compiler/src/Language/Bond/Codegen/Java/SerializationMethods.hs

```diff
@@ -115,10 +116,108 @@ writeMap java keyType valueType fieldName depth =
+            [lt|com.microsoft.bond.protocol.TaggedProtocolReader.ReadContainerResult|]
+
 deserialize_ProtocolWriter :: MappingContext -> Declaration -> Text
 deserialize_ProtocolWriter java declaration = [lt|
     @Override
     public void deserialize(com.microsoft.bond.protocol.TaggedProtocolReader reader) throws java.io.IOException {
         initSchema();
-    }
-|]
+
+        reader.readStructBegin();
+        #{newlineSepEnd 2 readField fields}
+        reader.readStructEnd();
+    }|]
+    where
+        fields = structFields declaration
+        readField field@Field {..} = [lt|
+        reader.readFieldBegin(this.#{readFieldResultMember});
+        #{readFieldValue java fieldType fieldName 0}
+        reader.readFieldEnd();|]
+
+readFieldValue :: MappingContext -> Type -> String -> Int -> Text
+readFieldValue java fieldType fieldName depth =
+    readValue java fieldType ("this." ++ fieldName) depth
+
+readValue :: MappingContext -> Type -> String -> Int -> Text
+readValue java fieldType varName depth = case fieldType of
+    BT_Int8                  -> [lt|#{varName} = reader.readInt8();|]
+    BT_Int16                 -> [lt|#{varName} = reader.readInt16();|]
+    BT_Int32                 -> [lt|#{varName} = reader.readInt32();|]
+    BT_Int64                 -> [lt|#{varName} = reader.readInt64();|]
+    BT_UInt8                 -> [lt|#{varName} = reader.readUInt8();|]
+    BT_UInt16                -> [lt|#{varName} = reader.readUInt16();|]
+    BT_UInt32                -> [lt|#{varName} = reader.readUInt32();|]
+    BT_UInt64                -> [lt|#{varName} = reader.readUInt64();|]
+    BT_Float                 -> [lt|#{varName} = reader.readFloat();|]
+    BT_Double                -> [lt|#{varName} = reader.readDouble();|]
+    BT_Bool                  -> [lt|#{varName} = reader.readBool();|]
+    BT_String                -> [lt|#{varName} = reader.readString();|]
+    BT_WString               -> [lt|#{varName} = reader.readWString();|]
+    BT_Blob                  -> [lt|#{varName} = reader.readBytes();|]
+    BT_Nullable e            -> readNullable java e varName depth
+    BT_List e                -> readSequence java e varName depth
+    BT_Vector e              -> readSequence java e varName depth
+    BT_Set e                 -> readSequence java e varName depth
+    BT_Map k v               -> readMap java k v varName depth
+    BT_UserDefined Enum {} _ ->
+        [lt|#{varName} = new #{getTypeName java fieldType}(reader.readInt32());|]
+    -- FIXME: Recursive types will cause infinite recursion.
+    BT_UserDefined _ _       ->
+        [lt|#{varName} = new #{getTypeName java fieldType}(); #{varName}.deserialize(reader);|]
+    _                        -> [lt|// FIXME: Not implemented.|]
+
+readNullable :: MappingContext -> Type -> String -> Int -> Text
+readNullable java elemType fieldName depth =
+    [lt|reader.readListBegin(#{readContainerResultMember});
+        if (#{readContainerResultMember}.count == 0) {
```

Comment:
The perils of putting up PRs before you can write tests. 🤔 

I would think that a nullable field with a count > 1 would be considered malformed data.